src/wrap_mempool.cpp

```diff
@@ -275,21 +293,232 @@ namespace
       { }
 
       const super::pointer_type data() const
-      { return ptr(); }
+      { return m_ptr; }
+
+      size_t size() const
+      {
+        return m_size;
+      }
   };
 
+  // }}}
 
 
+  // {{{ buffer_pool_allocate
 
-  pooled_buffer *device_pool_allocate(
-      std::shared_ptr<pyopencl::memory_pool<cl_allocator_base> > pool,
-      pyopencl::memory_pool<cl_allocator_base>::size_type sz)
+  pooled_buffer *buffer_pool_allocate(
+      std::shared_ptr<pyopencl::memory_pool<buffer_allocator_base> > pool,
+      pyopencl::memory_pool<buffer_allocator_base>::size_type sz)
   {
     return new pooled_buffer(pool, sz);
   }
 
+  // }}}
+
+
+#if PYOPENCL_CL_VERSION >= 0x2000
+
+  // {{{ svm allocator
+
+  // FIXME: Does this need deferred and immediate just like the buffer-level
+  // allocators? (I.e. can I tell whether I am out of memory just from allocations?)
+  struct svm_held_pointer
+  {
+    void *ptr;
+    pyopencl::command_queue_ref queue;
+  };
+
+
+  class svm_allocator
+  {
+    public:
+      typedef svm_held_pointer pointer_type;
+      typedef size_t size_type;
+
+    protected:
+      std::shared_ptr<pyopencl::context> m_context;
+      cl_uint m_alignment;
+      cl_svm_mem_flags m_flags;
+      pyopencl::command_queue_ref m_queue;
+
+
+    public:
+      svm_allocator(std::shared_ptr<pyopencl::context> const &ctx,
+          cl_uint alignment, cl_svm_mem_flags flags=CL_MEM_READ_WRITE,
+          pyopencl::command_queue *queue=nullptr)
+        : m_context(ctx), m_alignment(alignment), m_flags(flags)
+      {
+        if (queue)
+          m_queue.set(m_queue.data());
```

Comment:
Should this line be:

```suggestion
          m_queue.set(queue->data());
```

Reply:
Yep, thanks! I'd noticed this myself, too, but I didn't have a chance to push my work while I was traveling. 0cf016ff73bc3f7985b7ca227c4a8194e847ba0e has a memory pool that executes, but running DG examples with it gives incorrect results.